service/src/main/java/bio/terra/workspace/service/policy/flight/LinkSpendProfilePolicyAttributesStep.java

```diff
@@ -0,0 +1,124 @@
+package bio.terra.workspace.service.policy.flight;
+
+import bio.terra.policy.model.TpsComponent;
+import bio.terra.policy.model.TpsObjectType;
+import bio.terra.policy.model.TpsPaoConflict;
+import bio.terra.policy.model.TpsPaoGetResult;
+import bio.terra.policy.model.TpsPaoUpdateResult;
+import bio.terra.policy.model.TpsPolicyInputs;
+import bio.terra.policy.model.TpsUpdateMode;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.exception.InternalLogicException;
+import bio.terra.workspace.service.policy.TpsApiDispatch;
+import bio.terra.workspace.service.resource.exception.PolicyConflictException;
+import bio.terra.workspace.service.spendprofile.SpendProfileId;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.util.List;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LinkSpendProfilePolicyAttributesStep implements Step {
+  private static final Logger logger =
+      LoggerFactory.getLogger(LinkSpendProfilePolicyAttributesStep.class);
+
+  private final UUID workspaceId;
+  private final SpendProfileId spendProfileId;
+  private final TpsApiDispatch tpsApiDispatch;
+
+  public LinkSpendProfilePolicyAttributesStep(
+      UUID workspaceId, SpendProfileId spendProfileId, TpsApiDispatch tpsApiDispatch) {
+    this.workspaceId = workspaceId;
+    this.spendProfileId = spendProfileId;
+    this.tpsApiDispatch = tpsApiDispatch;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    UUID spendProfileUUID;
+    try {
+      spendProfileUUID = UUID.fromString(spendProfileId.getId());
+    } catch (IllegalArgumentException e) {
+      logger.info(
+          "non-UUID spend profile id provided, skipping spend profile and workspace PAO linking. [spendProfileId={}, workspaceId={}]",
+          spendProfileId,
+          workspaceId);
+      return StepResult.getStepResultSuccess();
+    }
+
+    // Create PAOs if they don't exist; catch TPS exceptions and retry
+    TpsPaoGetResult workspacePao;
+    try {
+      tpsApiDispatch.getOrCreatePao(
+          spendProfileUUID, TpsComponent.BPM, TpsObjectType.BILLING_PROFILE);
+      workspacePao =
+          tpsApiDispatch.getOrCreatePao(workspaceId, TpsComponent.WSM, TpsObjectType.WORKSPACE);
+    } catch (Exception ex) {
```

Comment:
Can this be more focused to the relevant APIExceptions from the clients instead of a blanket `catch (Exception ex)`?